src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -184,7 +184,11 @@ private function parseDefinition(\DOMElement $service, $file)
                 } elseif ($childService = $factory->getAttribute('service')) {
                     $class = new Reference($childService, ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE, false);
                 } else {
-                    $class = $factory->getAttribute('class');
+                    $class = $factory->getAttribute('class') ?: $definition->getClass();
```

Comment:
Right. It's not needed, but allow to provide a better exception message. See https://github.com/symfony/symfony/pull/20943#discussion_r92930104